genesyscloud/resource_genesyscloud_telephony_providers_edges_phonebasesettings.go

```diff
@@ -275,3 +333,100 @@ func PhoneBaseSettingsExporter() *resourceExporter.ResourceExporter {
 		JsonEncodeAttributes: []string{"properties"},
 	}
 }
+
+func buildSdkCapabilities(d *schema.ResourceData) *platformclientv2.Phonecapabilities {
+	if capabilities := d.Get("capabilities").([]interface{}); capabilities != nil {
+		sdkPhoneCapabilities := platformclientv2.Phonecapabilities{}
+		if len(capabilities) > 0 {
+			if _, ok := capabilities[0].(map[string]interface{}); !ok {
+				return nil
+			}
+			capabilitiesMap := capabilities[0].(map[string]interface{})
+
+			// Only set non-empty values.
+			provisions := capabilitiesMap["provisions"].(bool)
+			registers := capabilitiesMap["registers"].(bool)
+			dualRegisters := capabilitiesMap["dual_registers"].(bool)
+			var hardwareIdType string
+			if checkHardwareIdType := capabilitiesMap["hardware_id_type"].(string); len(checkHardwareIdType) > 0 {
+				hardwareIdType = checkHardwareIdType
+			}
+			allowReboot := capabilitiesMap["allow_reboot"].(bool)
+			noRebalance := capabilitiesMap["no_rebalance"].(bool)
+			noCloudProvisioning := capabilitiesMap["no_cloud_provisioning"].(bool)
+			mediaCodecs := make([]string, 0)
+			if checkMediaCodecs := capabilitiesMap["media_codecs"].([]interface{}); len(checkMediaCodecs) > 0 {
+				for _, codec := range checkMediaCodecs {
+					mediaCodecs = append(mediaCodecs, fmt.Sprintf("%v", codec))
+				}
+			}
+			cdm := capabilitiesMap["cdm"].(bool)
+
+			sdkPhoneCapabilities = platformclientv2.Phonecapabilities{
+				Provisions:          &provisions,
+				Registers:           &registers,
+				DualRegisters:       &dualRegisters,
+				HardwareIdType:      &hardwareIdType,
+				AllowReboot:         &allowReboot,
+				NoRebalance:         &noRebalance,
+				NoCloudProvisioning: &noCloudProvisioning,
+				MediaCodecs:         &mediaCodecs,
+				Cdm:                 &cdm,
+			}
+		}
+		return &sdkPhoneCapabilities
+	}
+	return nil
+}
+
+func flattenPhoneCapabilities(capabilities *platformclientv2.Phonecapabilities) []interface{} {
+	if capabilities == nil {
+		return nil
+	}
+
+	capabilitiesMap := make(map[string]interface{})
+	if capabilities.Provisions != nil {
```

Comment:
We should make a function with this checks and call it.  Otherwise we are bloating up the code with a bunch of conditionals

Reply:
@carnellj

hi John, so upon checking this file isn't part of the refactored resources in the PR. Some of them were modified as a side effect of publicizing resource utility functions for use in the new packages.

My question is can we do this in a separate ticket? The 3 refactored resources were the least amount of resources affected to implement the 'default site' feature and I kinda didn't want to overload it with not directly related changes.

The 3 resources that were refactored though did include improvements in using the 'resourcedata' util functions

Reply:
Hi Prince. I think John is just talking about using the resourcedata function `SetMapValueIfNotNil` here instead of implementing all of these if statements

Reply:
Hi Charlie, updated the code to use the resourcedata function